diff --git a/pkg/fab/comm/connector.go b/pkg/fab/comm/connector.go
index 07043a2b..9055a6d9 100644
--- a/pkg/fab/comm/connector.go
+++ b/pkg/fab/comm/connector.go
@@ -169,6 +169,10 @@ func (cc *CachingConnector) createConn(ctx context.Context, target string, opts
 	cc.lock.Lock()
 	defer cc.lock.Unlock()
 
+	if cc.janitorDone == nil {
+		return nil, errors.New("caching connector is closed")
+	}
+
 	cconn, ok := cc.loadConn(target)
 	if ok {
 		return cconn, nil
@@ -225,6 +229,11 @@ func (cc *CachingConnector) shutdownConn(cconn *cachedConn) {
 	cc.lock.Lock()
 	defer cc.lock.Unlock()
 
+	if cc.janitorDone == nil {
+		logger.Debug("Connector already closed")
+		return
+	}
+
 	logger.Debugf("connection was shutdown [%s]", cconn.target)
 	cc.conns.Delete(cconn.target)
 	delete(cc.index, cconn.conn)
