diff --git a/src/cmd/link/internal/ld/data.go b/src/cmd/link/internal/ld/data.go
index 4523323..817d8ff 100644
--- a/src/cmd/link/internal/ld/data.go
+++ b/src/cmd/link/internal/ld/data.go
@@ -415,6 +415,15 @@ func relocsym(ctxt *Link, s *Symbol) {
 					// DWARF info between the compiler and linker.
 					continue
 				}
+			} else if Buildmode == BuildmodePlugin && r.Sym.Name == "main.main" {
+				// A plugin has no main.main symbol, but can import
+				// parts of the runtime that use funcPC(main_main)
+				// to reference main.main. That's OK, it is defined
+				// in the binary that loaded the plugin.
+				//
+				// This special case can be removed when issue #17150
+				// is implemented.
+				r.Sym.Type = SDYNIMPORT
 			} else {
 				Errorf(s, "relocation target %s not defined", r.Sym.Name)
 				continue
